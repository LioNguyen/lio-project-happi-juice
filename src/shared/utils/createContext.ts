import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
