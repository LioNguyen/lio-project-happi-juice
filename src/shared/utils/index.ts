export * from './lazyLoad'
export * from './tailwind'
