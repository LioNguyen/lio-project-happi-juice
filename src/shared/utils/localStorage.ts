/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
