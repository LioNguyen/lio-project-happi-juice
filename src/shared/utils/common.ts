import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
