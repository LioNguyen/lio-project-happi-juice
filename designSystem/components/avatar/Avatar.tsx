import { type FC, type HTMLAttributes, memo, useState } from "react";

import { IComponentBase } from "@/shared/types";
import { cn } from "@/shared/utils";

interface IAvatarProps extends IComponentBase {
  src?: string;
  name: string;
  size?: number | string;
  imgProps?: Omit<HTMLAttributes<HTMLImageElement>, "src" | "alt" | "onClick">;
  onClick?: () => void;
}

const Avatar: FC<IAvatarProps> = ({ className, src, name, size = 40, imgProps, onClick }) => {
  const [isImageLoaded, setIsImageLoaded] = useState(false);
  const { className: imgClassName, ...restImgProps } = imgProps || {};

  const handleLoad = () => {
    setIsImageLoaded(true);
  };

  const getDisplayChars = (name: string) => {
    const words = name.split(" ").filter(Boolean);
    if (words.length >= 2) {
      return (words[0].charAt(0) + words[1].charAt(0)).toUpperCase();
    }
    return words[0].charAt(0).toUpperCase();
  };

  const displayChars = getDisplayChars(name);

  return (
    <div
      data-testid="avatar-test"
      className={cn(
        "avatar relative flex items-center justify-center rounded-full bg-background-secondary overflow-hidden",
        onClick ? "cursor-pointer" : "",
        className,
      )}
      style={{ width: size, height: size }}
      onClick={onClick}
    >
      <span
        className={cn("text-white font-medium", isImageLoaded ? "hidden" : "block")}
        style={{ fontSize: `${Number(size) * 0.4}px` }}
      >
        {displayChars}
      </span>
      <img
        className={cn("w-full h-full object-cover", imgClassName, isImageLoaded ? "block" : "hidden")}
        src={src}
        alt={name}
        onLoad={handleLoad}
        {...restImgProps}
      />
    </div>
  );
};

export default memo(Avatar);
export type { IAvatarProps };
