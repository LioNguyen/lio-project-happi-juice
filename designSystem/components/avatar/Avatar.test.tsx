import { fireEvent, screen } from "@testing-library/react";
import { ComponentPropsWithRef } from "react";
import { vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import Avatar from "./Avatar";

type AvatarPropsWithRef = ComponentPropsWithRef<typeof Avatar>;

const defaultProps = {
  name: "License Manager",
  src: "./avatar.jpg",
};

const renderComponent = (props: Partial<AvatarPropsWithRef>) => {
  return renderWithTheme(<Avatar {...defaultProps} {...props} />);
};

describe("Avatar test suite", () => {
  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderComponent({});
    expect(container).toMatchSnapshot();
  });

  it("should render correctly", () => {
    renderComponent({});
    const avatar = screen.getByTestId("avatar-test");
    expect(avatar).toBeInTheDocument();
  });

  it("should render class correctly", () => {
    const className = "custom-class";
    renderComponent({ className });

    const avatar = screen.getByTestId("avatar-test");
    expect(avatar).toHaveClass("avatar");
    expect(avatar).toHaveClass(
      "relative flex items-center justify-center rounded-full bg-background-secondary overflow-hidden",
    );
    expect(avatar).not.toHaveClass("cursor-pointer");
    expect(avatar).toHaveClass(className);
  });

  it("should show first character for single word name", () => {
    renderComponent({ name: "Admin" });
    const char = screen.getByText("A");
    expect(char).toBeInTheDocument();
    expect(char).toHaveClass("block");
  });

  it("should show first two characters for multi-word name", () => {
    renderComponent({ name: "License Manager" });
    const chars = screen.getByText("LM");
    expect(chars).toBeInTheDocument();
    expect(chars).toHaveClass("block");
  });

  it("should show image when src is provided and image loads successfully", () => {
    renderComponent({});
    const img = screen.getByAltText(defaultProps.name);

    fireEvent.load(img);

    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute("src", defaultProps.src);
    expect(screen.queryByText("LM")).toHaveClass("hidden");
  });

  it("should show name fallback when src is not provided", () => {
    renderComponent({ src: "" });

    const chars = screen.getByText("LM");
    expect(chars).toBeInTheDocument();
    expect(screen.queryByRole("img")).toHaveClass("hidden");
  });

  it("should render size correctly", () => {
    const size = 60;
    renderComponent({ size });

    const avatar = screen.getByTestId("avatar-test");
    expect(avatar).toHaveStyle({ width: `${size}px`, height: `${size}px` });
  });

  it("should handle onClick correctly", () => {
    const onClick = vi.fn();
    renderComponent({ onClick });

    const avatar = screen.getByTestId("avatar-test");
    fireEvent.click(avatar);

    expect(avatar).toHaveClass("cursor-pointer");
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("should render imgProps correctly when image is shown", () => {
    const imgProps = {
      className: "custom-img-class",
    };
    renderComponent({ imgProps });

    const img = screen.getByAltText(defaultProps.name);
    fireEvent.load(img);

    expect(img).toHaveClass(imgProps.className);
  });
});
