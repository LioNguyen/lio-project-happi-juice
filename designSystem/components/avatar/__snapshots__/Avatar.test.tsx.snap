// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar test suite > should match snapshot with default props 1`] = `
<div>
  <div
    class="avatar relative flex items-center justify-center rounded-full bg-background-secondary overflow-hidden"
    data-testid="avatar-test"
    style="width: 40px; height: 40px;"
  >
    <span
      class="text-white font-medium block"
      style="font-size: 16px;"
    >
      LM
    </span>
    <img
      alt="License Manager"
      class="w-full h-full object-cover hidden"
      src="./avatar.jpg"
    />
  </div>
</div>
`;
